### ISSUE TITLE
Switch from cookiecutter to copier

### ISSUE THREAD

**Issue by ruestefa** [MeteoSwiss-APN/mch-python-blueprint#31]
## TL;DR

I propose to switch from [cookiecutter](https://cookiecutter.readthedocs.io/en/stable/) to [copier](https://copier.readthedocs.io/en/stable/) because the latter enables projects to be easily updated when the underlying template changes.

_(I didn't propose this at the last meeting because I had not actually tried out updating a repo yet, but did so today for multiple repos ([template](https://github.com/ruestefa/python-project-template)), which worked beautifully and convinced me the benefits of `copier`.)_

## Cookiecutter

New project based on the blueprint are currently created with [cookiecutter](https://cookiecutter.readthedocs.io/en/stable/). The user is thereby prompted a few questions (such as the name of the project, email address, version etc.), and placeholders in the template files are replaced by the answers (leaving more complex conditionals aside). Once a project has been created, it is no longer dependent on cookiecutter, but when the template changes (e.g., an updated linter config, or a bugfix in a utility script), there is no straightforward way to update the project.

## Copier

[Copier](https://copier.readthedocs.io/en/stable/) works pretty much exactly like cookiecutter during project creation: The user is prompted questions, and based on the answers variables in the template files are replaced; even the syntax is mostly compatible (jinja templating). However, projects can stay connected to the template and easily updated when something in the template changes, as explained below. This prevents projects from drifting away from the template, and thereby away from each other and is a huge advantage over cookiecutter.

### Updating process

When updating a project to the underlying template, copier does two things:

- It asks the user the same questions that it already asked during creation of the project (defaulting to the latest answers).
- It checks the git history of the project for changes to the files that are part of the template, incl. whether files have been moved or deleted.

Based on the new answers (apart from the version number, most probably won't change frequently), copier creates an up-to-date version of the template, and then tries to apply the changes to that temporary project that is has detected in the git history of the project. The user then manually accepts or rejects the changes using standard git tools (`git add [-p]`, `git restore`, `git clean` etc.) and finally commits them.

All copier needs for a project to remain connected to a template is one file (`.copier-answers.yml`) in which it stores the last set of answers and the commit hash of the template repo at the point of the last update. Given the user always has last say by committing all changes manually, there is also little danger of messing anything up even in complex cases.

For more details, see the [docs](https://copier.readthedocs.io/en/latest/updating/).

## Next steps

- I could to show a small demo of the updating process at our next meeting.
- If accepted, I'd of course take over the tasks of implementation (little work) and documentation.

**Comment by ruestefa** [comment]
Meanwhile, I've gained more practical experience with `copier` with my own [templates](https://git.iac.ethz.ch/atmcirc/templates/). Some points to note:

- Most flags (notably `-f` and `-a`) belong to the main `copier` command, not the subcommands like `update`, and therefore have to precede the latter: `copier -a .copier-answers.yaml -f update`.
- `copier update` is fairly verbose regarding conflicts it encounters during the update/merge process and asks for permission to overwrite even in some cases where there is no net change. To avoid this, always use the force flag: `copier -f update`. The changes introduced by the update will have to be reviewed manually (with git commands etc.) anyway.
- It is feasible to create meta templates, which are used to create valid `copier` templates, by simply adding another layer of `tmpl/`, escaping the jinja directives etc.Real example from my templates:
  - `meta-python-project`: Meta version of a complete regular python project template (the equivalent of the blueprint).
  - `python-project`: Full template for regular Python projects. Depends only on `meta-python-project`.
  - `meta-python-f2py`: Adaptations necessary to compile `f2py` extension modules: Adapted `setup.py`, `CMakeLists.txt`, additional dependencies in `pyproject.toml`, `requirements.yml` and `dev-requirements.yml` etc.
  - `python-project-f2py`: Full template for Python projects with `f2py`. Depends on both `meta-python-project` and `python-project-f2py`.
